snuba/web/views.py

```diff
@@ -92,23 +96,42 @@ def check_clickhouse() -> bool:
     Checks if all the tables in all the enabled datasets exist in ClickHouse
     """
     try:
-        for name in get_enabled_dataset_names():
-            dataset = get_dataset(name)
-            for entity in dataset.get_all_entities():
-                for storage in entity.get_all_storages():
-                    clickhouse = storage.get_cluster().get_query_connection(
-                        ClickhouseClientSettings.QUERY
-                    )
-                    clickhouse_tables = clickhouse.execute("show tables")
-                    source = storage.get_schema()
-                    if isinstance(source, TableSchema):
-                        table_name = source.get_table_name()
-                        if (table_name,) not in clickhouse_tables:
-                            return False
+        datasets = [get_dataset(name) for name in get_enabled_dataset_names()]
+        entities = itertools.chain(
+            *[dataset.get_all_entities() for dataset in datasets]
+        )
+        storages = list(
+            itertools.chain(*[entity.get_all_storages() for entity in entities])
+        )
+
+        cluster_grouped_table_names: Dict[ConnectionId, List[str]] = defaultdict(list)
+        for storage in storages:
+            if isinstance(storage.get_schema(), TableSchema):
+                cluster = storage.get_cluster()
+                cluster_grouped_table_names[cluster.get_connection_id()].append(
+                    cast(TableSchema, storage.get_schema()).get_table_name()
+                )
+
+        # De-dupe clusters by host:port pairs
+        unique_clusters = {
+            storage.get_cluster().get_connection_id(): storage.get_cluster()
+            for storage in storages
+        }
+
+        for (cluster_key, cluster) in unique_clusters.items():
+            clickhouse = cluster.get_query_connection(ClickhouseClientSettings.QUERY)
+            clickhouse_tables = clickhouse.execute("show tables")
+            known_table_names = cluster_grouped_table_names[cluster_key]
+            logger.debug(f"checking for {known_table_names} on {cluster_key}")
+            for table in known_table_names:
```

Comment:
Sorry, I think we forgot one dimension.
On the same host:port pair you can have multiple databases, which will be on separate clusters and will require a separate `SHOW TABLES` command.
So doing one `SHOW TABLES` per host:port is not enough. it has to be per host:port:database